content/client-lib-development-guide/features.textile

```diff
@@ -1464,6 +1486,110 @@ enum StatsIntervalGranularity:
   DAY
   MONTH
 
+class DeviceDetails:
+  + getDeviceInfo() => io DeviceDetails
+  id: String
+  platform: DeviceType
+  type: DeviceType
+  clientId: String?
+  metadata: Dict<String, String>
+  updateTokenId: String
+  resetId(String)
+  resetUpdateToken() => io
+  push: DevicePushDetails
+
+class DevicePushDetails:
+  transportType: DevicePushTransportType
+  transportId: String
+  transportToken: String?
+  state: .Active | .Failing | .Failed
+  errorReason: ErrorInfo?
+  metadata: Dict<String, String>
+
+class Push:
+  publish(clientId: String, JsonObject) => io
+  publish(deviceId: String, JsonObject) => io
+  publish(transportType: String, transportId: String, transportTokenOrKey?: String) => io
+  admin: PushAdmin
+
+  // Only on platforms that support receiving notifications:
+
+  activate(DeviceDetails) => io ErrorInfo? // RSH2c
+  activateExternally(
+    DeviceDetails,
+    registerDevice: (ErrorInfo?, DeviceDetails?) -> io String?,
+    renewRegistration: (ErrorInfo?, DeviceDetails?) -> io String?
+  )
+  deactivate(deviceId: String) => io ErrorInfo?
+  deactivateExternally
+    deviceId: String,
+    deactivate: (ErrorInfo?, deviceId: String) -> io
+  )
+
+class PushAdmin:
+  deviceRegistrations: PushDeviceRegistrations
+  channelSubscriptions: PushChannelSubscriptions
+
+class JsonObject:
+  // Platform-dependant, typically a Dict-like object
+
+class PushDeviceRegistrations:
+  save(PushDeviceRegistration) => io
+  find(type: PushDeviceType, transportId: String) => io PushDeviceRegistration?
```

Comment:
I am finding the differentiation between find & get a bit obscure i.e. `find` returns one, whereas `get` multiple items.  To simplify things, why don't we have a set of params that can be used and have a single method `get` that is used for all queries such as:

```
get(queryParams: Params<String, String>)
```

So it can be used as follows:
```
get() => Page through all devices
get({ "clientId": "bob" }) => give me all devices for bob
get({ "deviceId": "xxx" }) => give me all devices for that device Id, which should only be one
get({ "transportPlatform": "ios", "formFactor": "mobile" }) => assuming we supported this in the REST API, they could get all iOS devices with a mobile form factor
```

I don't see the value in typing these queries to be honest.

Same applies for `remove`, why not only allow remove by `deviceId` which is unique, and if a user needs to look up a `transportId` and delete it, then let them do that, so `remove(deviceId: string) | remove(device: DeviceDetails) | remove(Array<string | DeviceDetails>)`.

I think taking this approach would simplify things immensely and allow us to add support for new query types more easily too.